File: src/tpuff/commands/scan.py
"""Scan command for tpuff CLI — extract unique field values from a namespace."""

import json
import sys

import click
from rich.console import Console

from tpuff.client import get_namespace
from tpuff.utils.debug import debug_log
from tpuff.utils.output import is_plain

# stderr console so progress doesn't pollute JSON output
err_console = Console(stderr=True)


@click.command("scan", context_settings={"help_option_names": ["-h", "--help"]})
@click.option("-n", "--namespace", required=True, help="Namespace to scan")
@click.option("--field", required=True, help="Field name to extract unique values from")
@click.option("-r", "--region", help="Override the region (e.g., aws-us-east-1, gcp-us-central1)")
@click.option("--page-size", default=1000, type=int, help="Batch size per query (default: 1000)")
@click.pass_context
def scan(
    ctx: click.Context,
    namespace: str,
    field: str,
    region: str | None,
    page_size: int,
) -> None:
    """Scan a namespace and extract all unique values of a field."""
    plain = is_plain(ctx)

    try:
        ns = get_namespace(namespace, region)

        # Get total document count from metadata for progress bar
        metadata = ns.metadata()
        meta_dict = metadata.model_dump() if hasattr(metadata, "model_dump") else {}
        total_docs = meta_dict.get("approx_row_count") or None

        unique_values: set[str] = set()
        last_id: str | None = None
        total_scanned = 0

        if not plain:
            from rich.progress import (
                BarColumn,
                MofNCompleteColumn,
                Progress,
                SpinnerColumn,
                TextColumn,
            )

            progress = Progress(
                SpinnerColumn(),
                TextColumn("[progress.description]{task.description}"),
                BarColumn(bar_width=None),
                MofNCompleteColumn(),
                console=err_console,
            )
            task = progress.add_task("Scanning...", total=total_docs)
            progress.start()

        while True:
            query_params: dict = {
                "rank_by": ["id", "asc"],
                "include_attributes": [field],
                "top_k": page_size,
            }

            if last_id is not None:
                query_params["filters"] = ["id", "Gt", last_id]

            debug_log("Scan query", query_params)
            result = ns.query(**query_params)
            rows = result.rows if hasattr(result, "rows") else []

            for row in rows:
                if hasattr(row, "model_dump"):
                    row_dict = row.model_dump()
                else:
                    row_dict = {"id": getattr(row, "id", None)}

                value = row_dict.get(field)
                if value is not None:
                    unique_values.add(value)

                last_id = row_dict.get("id", getattr(row, "id", None))

            total_scanned += len(rows)

            if not plain:
                progress.update(
                    task,
                    completed=total_scanned,
                    description=f"{len(unique_values)} unique values",
                )

            debug_log("Scan page", {
                "rows": len(rows),
                "total_scanned": total_scanned,
                "unique_values": len(unique_values),
                "last_id": last_id,
            })

            if len(rows) < page_size:
                break

        if not plain:
            progress.stop()
            err_console.print(
                f"[green]Done.[/green] Scanned {total_scanned} documents, "
                f"found {len(unique_values)} unique values."
            )

        # Output the unique values as a sorted JSON array to stdout
        click.echo(json.dumps(sorted(unique_values)))

    except Exception as e:
        if not plain:
            try:
                progress.stop()
            except Exception:
                pass
        err_console.print(f"[red]Error: {e}[/red]")
        sys.exit(1)

File: cmd/scan.go
package cmd

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"sort"

	"github.com/hev/tpuff/internal/client"
	"github.com/hev/tpuff/internal/debug"
	"github.com/hev/tpuff/internal/output"
	"github.com/spf13/cobra"
	"github.com/turbopuffer/turbopuffer-go"
)

var scanCmd = &cobra.Command{
	Use:   "scan",
	Short: "Scan a namespace and extract all unique values of a field",
	RunE:  runScan,
}

func init() {
	scanCmd.Flags().StringP("namespace", "n", "", "Namespace to scan")
	scanCmd.Flags().String("field", "", "Field name to extract unique values from")
	scanCmd.Flags().StringP("region", "r", "", "Override the region")
	scanCmd.Flags().Int("page-size", 1000, "Batch size per query")
	_ = scanCmd.MarkFlagRequired("namespace")
	_ = scanCmd.MarkFlagRequired("field")
	rootCmd.AddCommand(scanCmd)
}

func runScan(cmd *cobra.Command, args []string) error {
	namespace, _ := cmd.Flags().GetString("namespace")
	field, _ := cmd.Flags().GetString("field")
	region, _ := cmd.Flags().GetString("region")
	pageSize, _ := cmd.Flags().GetInt("page-size")
	ctx := context.Background()

	ns, err := client.GetNamespace(namespace, region)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
		os.Exit(1)
	}

	// Get total count for progress
	md, err := ns.Metadata(ctx, turbopuffer.NamespaceMetadataParams{})
	totalDocs := int64(0)
	if err == nil {
		totalDocs = md.ApproxRowCount
	}

	uniqueValues := make(map[string]bool)
	var lastID string
	totalScanned := 0

	for {
		queryParams := turbopuffer.NamespaceQueryParams{
			RankBy:            turbopuffer.NewRankByAttribute("id", turbopuffer.RankByAttributeOrderAsc),
			IncludeAttributes: turbopuffer.IncludeAttributesParam{StringArray: []string{field}},
			TopK:              turbopuffer.Int(int64(pageSize)),
		}

		if lastID != "" {
			queryParams.Filters = turbopuffer.NewFilterGt("id", lastID)
		}

		debug.Log("Scan query", map[string]any{
			"last_id":   lastID,
			"page_size": pageSize,
		})

		result, err := ns.Query(ctx, queryParams)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
			os.Exit(1)
		}

		for _, row := range result.Rows {
			if v, ok := row[field]; ok && v != nil {
				uniqueValues[fmt.Sprintf("%v", v)] = true
			}
			if id, ok := row["id"]; ok {
				lastID = fmt.Sprintf("%v", id)
			}
		}

		totalScanned += len(result.Rows)

		if !output.IsPlain() {
			if totalDocs > 0 {
				fmt.Fprintf(os.Stderr, "\r%d unique values — scanned %d/%d documents",
					len(uniqueValues), totalScanned, totalDocs)
			} else {
				fmt.Fprintf(os.Stderr, "\r%d unique values — scanned %d documents",
					len(uniqueValues), totalScanned)
			}
		}

		debug.Log("Scan page", map[string]any{
			"rows":          len(result.Rows),
			"total_scanned": totalScanned,
			"unique_values": len(uniqueValues),
			"last_id":       lastID,
		})

		if len(result.Rows) < pageSize {
			break
		}
	}

	if !output.IsPlain() {
		fmt.Fprintf(os.Stderr, "\nDone. Scanned %d documents, found %d unique values.\n",
			totalScanned, len(uniqueValues))
	}

	// Output sorted unique values as JSON array
	sorted := make([]string, 0, len(uniqueValues))
	for v := range uniqueValues {
		sorted = append(sorted, v)
	}
	sort.Strings(sorted)

	b, _ := json.Marshal(sorted)
	fmt.Println(string(b))
	return nil
}
